---
title: "Brachypodium QTL scans"
author: "Hugo Tavares"
date: "2019-06-17"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message = FALSE, warning = FALSE}
# load libraries
library(qtl2)
library(tidyverse)
library(patchwork)
library(qtl2helper) # https://github.com/tavareshugo/qtl2helper

# change ggplot2 defaults
theme_set(theme_bw() + theme(text = element_text(size = 12)))
```


## Summary

This documents drafts the analysis routine used for QTL mapping. 
The genotype data is imported using the `R/qtl2` package, which is used to infer genotype probabilities at evenly spaced (pseudo-)markers from the SNP data of Cui et al 2012.

We then use two strategies for QTL mapping:

* A simple QTL scan using averaged trait values.
* Using a GxE mixed model for the QTL scan, using the non-summarised data.


## Prepare `R/qtl2` data

#### Download genotype data

The genotype data from Cui et al 2012 was cleaned and formatted for analysis with the `R/qtl2` package. 
See [this repository](https://github.com/tavareshugo/qtl2data/tree/master/brachy_Cui2012) for further details about pre-processing of these data. 
The data can be downloaded from this repository as a compressed _.zip_ file, and read with the `qtl2::read_cross2()` function.

```{r}
# prepare directory structure
if(!file.exists("../../data/external/")) dir.create("../../data/external", recursive = TRUE)

# Download genotype data
if(!file.exists("../../data/external/Cui2012_qtl2cross.zip")){
  download.file("https://github.com/tavareshugo/qtl2data/raw/master/brachy_Cui2012/Cui2012_qtl2cross.zip",
                destfile = "../../data/external/Cui2012_qtl2cross.zip",
                method = "wget")
}

# read data as a cross2 object
ril_cross <- read_cross2("../../data/external/Cui2012_qtl2cross.zip")
```


#### Prepare trait data

We read the clean (processed) data and summarise per genotype/nitrate treatment

```{r}
# read individual data
ril_phenotypes <- read_csv("../../data/processed/ril_phenotypes/ril_phenotypic_data_individual.csv",
                           col_types = cols(
                             .default = col_double(),
                             sheet = col_character(),
                             identifier = col_character(),
                             block = col_character(),
                             shelf = col_character(),
                             nitrate_level = col_factor(levels = c("LN", "HN")),
                             id = col_character(),
                             batch = col_factor(),
                             position_grid_location = col_character(),
                             ril_id = col_character(),
                             sow_date = col_date(format = "%Y-%m-%d"),
                             z68_date = col_date(format = "%Y-%m-%d"),
                             senescence_date = col_date(format = "%Y-%m-%d")
                           ))
# 
# # Simple summary of data
# pheno <- ril_phenotypes %>% 
#   # retain only RILs
#   filter(str_detect(ril_id, "RIL")) %>% 
#   # summarise per nitrate and RIL
#   group_by(nitrate_level, ril_id) %>% 
#   summarise_at(vars(z68_shoots, z68_greenness, senescence_height, seed_weight_g, 
#                     lifespan, flowering_senescence_interval, flowering_time),
#                .funs = list(mean = ~ mean(., na.rm = TRUE),
#                             n = ~ sum(!is.na(.)),
#                             sd = ~ sd(., na.rm = TRUE))) %>% 
#   # remove grouping
#   ungroup()
# 
# # Summarised data is also available from this file
# # but there are some differences from above summary, 
# # possibly because the table below is summarised per batch and then per ril/nitrate?
# # pheno <- read_csv("../../data/processed/ril_phenotypes/2019_05_28_averages_RILs.csv")

# read summarised phenotype data
pheno <- read_csv("../../data/processed/ril_phenotypes/ril_phenotypic_data_averaged.csv",
                           col_types = cols(
                             .default = col_double(),
                             nitrate_level = col_factor(levels = c("LN", "HN")),
                             ril_id = col_character()))

# Get HN and LN into separate columns as well as plasticity
pheno_summary <- pheno %>%
  filter(str_detect(ril_id, "RIL")) %>% 
  select(ril_id, nitrate_level, matches("_mean")) %>%
  drop_na() %>%
  group_by(ril_id) %>%
  summarise_at(vars(matches("_mean")),
               .funs = list(hn = ~ .[nitrate_level == "HN"],
                            ln = ~ .[nitrate_level == "LN"],
                            plas = ~ .[nitrate_level == "HN"] - .[nitrate_level == "LN"])) %>%
  # scale the data (this is probably not really needed...)
  mutate_at(vars(matches("_mean")),
            list(~ (. - mean(.))/sd(.)))
```


These data can now be added to our `cross2` object, to ensure consistency between 
the genotypes on our data and the ones with genotype information:

```{r}
ril_cross <- add_pheno(ril_cross, pheno_summary)
```


#### Prepare map data

Finally, we impute genotypes at even intervals (0.5cM), and then calculate genotype probabilities from the SNP data.

```{r}
# insert pseudo-markers for evenly spaced scan
geno_map <- insert_pseudomarkers(ril_cross$gmap, step = 0.5)

# calculate genotype probabilities - assume 1% error rate
geno_prob <- calc_genoprob(ril_cross, geno_map, error_prob = 0.01, cores = 3)
```



## Simple QTL scan

The simple QTL scan is straighforward, using `R/qtl2` functionality:

```{r}
# Run scan
pheno_scan <- scan1(geno_prob, ril_cross$pheno, cores = 2)

# Run permutations (takes a while)
pheno_perm <- scan1perm(geno_prob, ril_cross$pheno, n_perm = 500, cores = 3)
```


#### Visualise results

Visualising the results can be done quickly using `R/qtl2` functions.

```{r}
# QTL scan for shoot branching
plot_scan1(pheno_scan, geno_map, lodcolumn = "z68_shoots_mean_hn")

# add genome-wide significance threshold
abline(h = summary(pheno_perm)[, "z68_shoots_mean_hn"], lty = 2)
```

But it might be more convenient to "tidy" the data so that one can make custom plots more easily (e.g. with `ggplot2`).

```{r}
pheno_scan_df <- pheno_scan %>% 
  # tidy (from qtl2helper package)
  tidy(map = geno_map) %>% 
  # add nitrate variable based on pheno column
  mutate(nitrate = toupper(str_remove(pheno, ".*_")))

# Tidy permutations
pheno_perm_df <- pheno_perm %>% 
  tidy(alpha = c(0.1, 0.05, 0.01)) %>% 
  # create variable with nitrate treatment
  mutate(nitrate = toupper(str_remove(pheno, ".*_")))

# get peaks
# using threshold of 3, as that seems to be consistent across traits from permutation test
pheno_peak_df <- find_peaks(pheno_scan, geno_map, threshold = 3, drop = 1) %>% 
  rename(pheno = lodcolumn, chrom = chr) %>% 
  mutate(nitrate = toupper(str_remove(pheno, ".*_")))
```

Here is the plot now:

```{r}
pheno_scan_df %>% 
  filter(str_detect(pheno, "z68_shoots_mean")) %>% 
  ggplot(aes(pos, LOD)) +
  geom_line() +
  geom_hline(data = pheno_perm_df %>% filter(str_detect(pheno, "z68_shoots_mean") & alpha == "0.05"),
             aes(yintercept = threshold), linetype = 2) +
  geom_segment(data = pheno_peak_df %>% filter(str_detect(pheno, "z68_shoots_mean")),
               aes(x = ci_lo, xend = ci_hi, y = -0.1, yend = -0.1), 
               colour = "brown", size = 3) +
  facet_grid(nitrate ~ chrom, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(0, 500, 100)) +
  labs(x = "Position (cM)") +
  theme(panel.grid = element_blank())
```

Let's investigate that marker a bit more. First let's identify which marker is the top one. 

```{r}
# identify top marker
top_snp <- pheno_scan %>% 
  as_tibble(rownames = "marker") %>%
  top_n(1, z68_shoots_mean_hn) %>% 
  pull(marker)
```

Now, we can extract the genotype probabilities for that marker and add them to our table of phenotypes:

```{r}
# create temporary table with genotypes and phenotypes
temp <- pull_genoprobpos(geno_prob, top_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id")

temp %>% 
  ggplot(aes(cut(aa, c(0, 0.25, 0.75, 1)), z68_shoots_mean_hn)) +
  ggbeeswarm::geom_quasirandom()

# plot - storing it in a variable because will re-use it later
p1 <- temp %>% 
  ggplot(aes(flowering_time_mean_hn, z68_shoots_mean_hn)) +
  geom_point(aes(colour = aa), size = 2) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(colour = "Probability\nBd21", title = "Simple scan", tag = "A")
p1
```

### Run scan `z68_shoots` with `flowering_time` as covariate

Flowering and shoots are correlated, thus some QTL in common:

```{r}
p1 <- pheno_summary %>% 
  ggplot(aes(z68_shoots_mean_hn, flowering_time_mean_hn)) +
  geom_point() + 
  labs(x = "Shoots", y = "Flowering", title = "HN")

p2 <- pheno_summary %>% 
  ggplot(aes(z68_shoots_mean_ln, flowering_time_mean_ln)) +
  geom_point() + 
  labs(x = "Shoots", y = "Flowering", title = "LN")

p1 + p2
```


```{r}
p1 <- pheno_scan_df %>% 
  filter(str_detect(pheno, "z68_shoots_mean")) %>% 
  ggplot(aes(pos, LOD)) +
  geom_line() +
  geom_hline(data = pheno_perm_df %>% filter(str_detect(pheno, "z68_shoots_mean") & alpha == "0.05"),
             aes(yintercept = threshold), linetype = 2) +
  geom_segment(data = pheno_peak_df %>% filter(str_detect(pheno, "z68_shoots_mean")),
               aes(x = ci_lo, xend = ci_hi, y = -0.1, yend = -0.1), 
               colour = "brown", size = 3) +
  facet_grid(nitrate ~ chrom, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(0, 500, 100)) +
  labs(x = "Position (cM)", title = "Shoots") +
  theme(panel.grid = element_blank())

p2 <- pheno_scan_df %>% 
  filter(str_detect(pheno, "flowering_time_mean")) %>% 
  ggplot(aes(pos, LOD)) +
  geom_line() +
  geom_hline(data = pheno_perm_df %>% filter(str_detect(pheno, "flowering_time_mean") & alpha == "0.05"),
             aes(yintercept = threshold), linetype = 2) +
  geom_segment(data = pheno_peak_df %>% filter(str_detect(pheno, "flowering_time_mean")),
               aes(x = ci_lo, xend = ci_hi, y = -0.1, yend = -0.1), 
               colour = "brown", size = 3) +
  facet_grid(nitrate ~ chrom, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(0, 500, 100)) +
  labs(x = "Position (cM)", title = "Flowering") +
  theme(panel.grid = element_blank())

p1 + p2 + plot_layout(ncol = 1)
```

We can run QTL scan adding flowering as a covariate:

```{r}
# Run scan
pheno_scan_covar <- scan1(geno_prob, 
                          pheno = ril_cross$pheno[, c("z68_shoots_mean_hn"), drop = FALSE], 
                          addcovar = ril_cross$pheno[, c("flowering_time_mean_hn"), drop = FALSE], 
                          cores = 3) %>% 
  tidy(map = geno_map) %>% 
  mutate(nitrate = toupper(str_remove(pheno, ".*_")),
         covariate = TRUE)

# add to original scan
pheno_scan_df <- pheno_scan_df %>% 
  mutate(covariate = FALSE) %>% 
  bind_rows(pheno_scan_covar)

# Run permutations with covariate
pheno_covar_perm <- scan1perm(geno_prob, 
                          pheno = ril_cross$pheno[, c("z68_shoots_mean_hn"), drop = FALSE], 
                          addcovar = ril_cross$pheno[, c("flowering_time_mean_hn"), drop = FALSE], 
                          n_perm = 500,
                          cores = 3) %>% 
  tidy(alpha = c(0.1, 0.05, 0.01)) %>% 
  mutate(nitrate = toupper(str_remove(pheno, ".*_")))
```

```{r}
pheno_scan_df %>% 
  filter(str_detect(pheno, "z68_shoots") & nitrate == "HN") %>% 
  ggplot(aes(pos, LOD)) +
  geom_line(aes(colour = covariate)) +
  geom_hline(data = pheno_perm_df %>% filter(str_detect(pheno, "z68_shoots_mean_hn") & alpha == "0.05"),
             aes(yintercept = threshold), linetype = 2) +
    geom_hline(data = pheno_covar_perm %>% filter(alpha == "0.05"),
             aes(yintercept = threshold), linetype = 2) +
  facet_grid(nitrate ~ chrom, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(0, 500, 100)) +
  scale_colour_manual(values = c("black", "brown")) +
  labs(x = "Position (cM)", title = "Shoots") +
  theme(panel.grid = element_blank())
```

Let's investigate that marker a bit more. First let's identify which marker is the top one. 

```{r}
# identify top marker
top_snp_covar <- pheno_scan_covar %>% 
  top_n(1, LOD) %>% 
  pull(marker)
```

Now, we can extract the genotype probabilities for that marker and add them to our table of phenotypes:

```{r}
# create temporary table with genotypes and phenotypes
temp <- pull_genoprobpos(geno_prob, top_snp_covar) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id")

# plot - storing it in a variable because will re-use it later
p2 <- temp %>% 
  ggplot(aes(flowering_time_mean_hn, z68_shoots_mean_hn)) +
  geom_point(aes(colour = aa), size = 2) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(colour = "Probability\nBd21", title = "Scan with flowering covariate", tag = "B")
p2
```

From this we can see that on average *Bd21* genotypes have more branches, for a given flowering time. 
We can super-impose the regression lines from the QTL model. 
I will fit the model myself, but this could also be achieved with the `qtl2::fit1()` function.

```{r}
# Fit no interaction covariate model
covar_qtl_m1 <- lm(z68_shoots_mean_hn ~ 0 + aa + bb + flowering_time_mean_hn, data = temp)

summary(covar_qtl_m1)
```

So now we can overlay the model prediction on the previous graph:

```{r}
p2 +
  geom_abline(intercept = coef(covar_qtl_m1)["aa"], 
              slope = coef(covar_qtl_m1)["flowering_time_mean_hn"]) +
  geom_abline(intercept = coef(covar_qtl_m1)["bb"], 
              slope = coef(covar_qtl_m1)["flowering_time_mean_hn"])
```

As it is, this model expects that the effect of flowering is the same across genotypes, but it seems from the graph that actually an interaction model (where the slopes can vary) might be better? 
Let's try that!

```{r}
# Fit interaction covariate model
covar_qtl_m2 <- lm(z68_shoots_mean_hn ~ 0 + aa + bb + aa : flowering_time_mean_hn + bb : flowering_time_mean_hn,
                   data = temp)

summary(covar_qtl_m2)
```

Let's compare the two models:

```{r}
# Analysis of variance
anova(covar_qtl_m1, covar_qtl_m2)

# Difference in information criteria
AIC(covar_qtl_m1) - AIC(covar_qtl_m2)
```

Perhaps the complex model is marginally better, let's super-impose its fit on the graph:

```{r}
p2 +
  geom_abline(intercept = coef(covar_qtl_m1)["aa"], 
              slope = coef(covar_qtl_m1)["flowering_time_mean_hn"]) +
  geom_abline(intercept = coef(covar_qtl_m1)["bb"], 
              slope = coef(covar_qtl_m1)["flowering_time_mean_hn"]) +
  geom_abline(intercept = coef(covar_qtl_m2)["aa"], 
              slope = coef(covar_qtl_m2)["aa:flowering_time_mean_hn"], 
              linetype = 2) +
  geom_abline(intercept = coef(covar_qtl_m2)["bb"], 
              slope = coef(covar_qtl_m2)["bb:flowering_time_mean_hn"],
              linetype = 2)
```

In conclusion: this might be an interesting (if minor) QTL that explains branchiness conditional on flowering time. 



## Effect sizes

```{r}
pheno_scan_df %>% 
  filter(pheno %in% c("z68_shoots_mean_hn", "flowering_time_mean_hn", "flowering_time_mean_ln")) %>% 
  group_by(pheno, covariate) %>% 
  top_n(1, LOD)
```

```{r}
# Shoot and Flowering SNP on HN
target_snp <- "BD3260_5"

p1 <- pull_genoprobpos(geno_prob, target_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id") %>% 
  ggplot(aes(cut(aa, breaks = c(0, 0.25, 0.75, 1)), z68_shoots_mean_hn)) +
  ggbeeswarm::geom_quasirandom(aes(colour = aa), show.legend = FALSE) +
  geom_pointrange(stat = "summary", fun.data = "mean_cl_boot", colour = "brown", size = 1) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(x = "Probability Bd21", y = "Effect size", title = "Shoots HN")

p2 <- pull_genoprobpos(geno_prob, target_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id") %>% 
  ggplot(aes(flowering_time_mean_hn, z68_shoots_mean_hn)) +
  geom_point(aes(colour = aa), size = 2) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(x = "Flowering HN", y = "Shoots HN", colour = "Probability Bd21")

p1 + p2
```

```{r}
# Shoot SNP using flowering as covariate
target_snp <- "c5.loc137.5"

p1 <- pull_genoprobpos(geno_prob, target_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id") %>% 
  ggplot(aes(cut(aa, breaks = c(0, 0.25, 0.75, 1)), z68_shoots_mean_hn)) +
  ggbeeswarm::geom_quasirandom(aes(colour = aa), show.legend = FALSE) +
  geom_pointrange(stat = "summary", fun.data = "mean_cl_boot", colour = "brown", size = 1) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(x = "Probability Bd21", y = "Effect size", title = "Shoots HN")

p2 <- pull_genoprobpos(geno_prob, target_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id") %>% 
  ggplot(aes(flowering_time_mean_hn, z68_shoots_mean_hn)) +
  geom_point(aes(colour = aa), size = 2) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(x = "Flowering HN", y = "Shoots HN", colour = "Probability Bd21")

p1 + p2
```

Here, I try to plot all genotype combinations. 
But in a way, conditioning on flowering is the same as conditioning on the main QTL (because they are correlated). So this plot reveals a similar thing as above. 

```{r}
shoot_effects <- pull_genoprobpos(geno_prob, "BD3260_5") %>% 
  as_tibble(rownames = "ril_id") %>% 
  select(ril_id, peak1 = aa)

shoot_effects <- pull_genoprobpos(geno_prob, "c5.loc137.5") %>% 
  as_tibble(rownames = "ril_id") %>% 
  select(ril_id, peak2 = aa) %>% 
  inner_join(shoot_effects, by = "ril_id")

shoot_effects <- shoot_effects %>% 
  inner_join(pheno_summary, by = "ril_id")

shoot_effects %>% 
  mutate(peak1 = case_when(peak1 < 0.25 ~ "Bd3-1", 
                           peak1 > 0.75 ~ "BD21",
                           TRUE ~ as.character(NA)),
         peak2 = case_when(peak2 < 0.25 ~ "Bd3-1", 
                           peak2 > 0.75 ~ "BD21",
                           TRUE ~ as.character(NA))) %>% 
  drop_na() %>% 
  mutate(genotype = paste(peak1, peak2, sep = " / ")) %>% 
  mutate(genotype = fct_reorder(genotype, z68_shoots_mean_hn)) %>% 
  ggplot(aes(genotype, z68_shoots_mean_hn)) +
  ggbeeswarm::geom_quasirandom(size = 2) +
  geom_pointrange(stat = "summary", fun.data = "mean_cl_boot", colour = "brown", size = 1) +
  labs(x = "Genotype", y = "Shoots HN")
```



```{r}
# SNP that pops up on LN
target_snp <- "BD0635_1"

p1 <- pull_genoprobpos(geno_prob, target_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id") %>% 
  ggplot(aes(cut(aa, breaks = c(0, 0.25, 0.75, 1)), flowering_time_mean_ln)) +
  ggbeeswarm::geom_quasirandom(aes(colour = aa), show.legend = FALSE) +
  geom_pointrange(stat = "summary", fun.data = "mean_cl_boot", colour = "brown", size = 1) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(x = "Probability Bd21", y = "Effect size", title = "Flowering LN")

p2 <- pull_genoprobpos(geno_prob, target_snp) %>% 
  as_tibble(rownames = "ril_id") %>% 
  inner_join(pheno_summary, by = "ril_id") %>% 
  ggplot(aes(flowering_time_mean_ln, flowering_time_mean_hn)) +
  geom_point(aes(colour = aa), size = 2) +
  scale_colour_gradient2(limits = c(0, 1), low = "#762a83", high = "#1b7837", mid = "grey", midpoint = 0.5) +
  labs(x = "Flowering LN", y = "Flowering HN", colour = "Probability Bd21")

p1 + p2
```

It seems like there might be two outliers there. 
In particular, the one that falls off the diagonal. 
In the mixed models, it was shrunk towards the diagonal, but there it was not and may have big influence in the model. 

Re-running the scan without that individual doesn't make a difference:

```{r}
pheno_summary %>% top_n(2, flowering_time_mean_ln)

# re-run the scan without this individual
temp <- ril_cross$pheno[pheno_summary$ril_id[!(pheno_summary$ril_id %in% c("RIL130", "RIL098"))], ]

scan1(geno_prob, temp) %>% 
  plot(map = geno_map, lodcolumn = "flowering_time_mean_ln")
```

If we fit the mixed model, we can see that the two nitrate treatments should be identical:

```{r}
library(lme4)
model_data <- ril_phenotypes %>% 
  filter(str_detect(ril_id, "RIL")) %>% 
  # retain only lines that have data in both nitrate treatments
  group_by(ril_id) %>% 
  filter(n_distinct(nitrate_level) == 2) %>% 
  ungroup()

fit1 <- lmer(flowering_time ~ nitrate_level  + (nitrate_level|ril_id) + (nitrate_level|batch),
     data = model_data,
     control = lmerControl(optimizer ="Nelder_Mead"))

fit2 <- lm(flowering_time ~ nitrate_level*ril_id,
     data = model_data)

temp <- model_data %>% 
  filter(str_detect(ril_id, "RIL")) %>% 
  distinct(ril_id, nitrate_level) %>% 
  mutate(batch = "a") %>% 
  mutate(pred = predict(fit1, newdata = ., allow.new.levels = TRUE),
         mean = predict(fit2, newdata = .)) %>% 
  pivot_wider(names_from = nitrate_level, values_from = c(pred, mean)) %>% 
  select(-batch) %>% 
  column_to_rownames("ril_id") %>% as.matrix()

scan1(geno_prob, temp) %>% 
  plot(geno_map, lodcolumn = "mean_LN")
scan1(geno_prob, temp) %>% 
  plot(geno_map, lodcolumn = "mean_HN", add = TRUE, col = "brown")
```


Investigate whether the signal on Chr5 for flowering are two separate QTL:

```{r}
scan1(geno_prob, 
      pheno = ril_cross$pheno[, c("flowering_time_mean_hn"), drop = FALSE], 
      addcovar = pull_genoprobpos(geno_prob, "BD3660_1")[, 1, drop = FALSE], 
      cores = 2) %>% 
  plot(geno_map, ylim = c(0, 4))

PEAK1 <- pull_genoprobpos(geno_prob, "BD3660_1")[, 1]
PEAK2 <- pull_genoprobpos(geno_prob, "c5.loc83.5")[, 1]
PHEN <- ril_cross$pheno[, c("flowering_time_mean_hn")]

fit_both <- lm(PHEN ~ 1 + PEAK1[names(PHEN)] + PEAK2[names(PHEN)])
fit_one <- lm(PHEN ~ 1 + PEAK1[names(PHEN)])
anova(fit_both, fit_one)
summary(fit_both)$adj.r.squared
summary(fit_one)$adj.r.squared
```


```{r}
flower_effects <- pull_genoprobpos(geno_prob, "BD3660_1") %>% 
  as_tibble(rownames = "ril_id") %>% 
  select(ril_id, peak1 = aa)

flower_effects <- pull_genoprobpos(geno_prob, "c5.loc83.5") %>% 
  as_tibble(rownames = "ril_id") %>% 
  select(ril_id, peak2 = aa) %>% 
  full_join(flower_effects, by = "ril_id")

flower_effects <- flower_effects %>% 
  full_join(pheno_summary, by = "ril_id")

flower_effects %>% 
  mutate(peak1 = case_when(peak1 < 0.25 ~ "Bd3-1", 
                           peak1 > 0.75 ~ "BD21",
                           TRUE ~ as.character(NA)),
         peak2 = case_when(peak2 < 0.25 ~ "Bd3-1", 
                           peak2 > 0.75 ~ "BD21",
                           TRUE ~ as.character(NA))) %>% 
  drop_na() %>% 
  mutate(genotype = paste(peak1, peak2, sep = " / ")) %>% 
  mutate(genotype = fct_reorder(genotype, flowering_time_mean_hn)) %>% 
  ggplot(aes(genotype, flowering_time_mean_hn)) +
  ggbeeswarm::geom_quasirandom(size = 2) +
  geom_point(stat = "summary", fun.y = "mean", colour = "brown", size = 3) +
  labs(x = "Genotype", y = "Flowering HN")
```


## Height

```{r}
pheno_scan_df %>% 
  filter(str_detect(pheno, "height")) %>% 
  ggplot(aes(pos, LOD)) +
  geom_line() +
  geom_hline(data = pheno_perm_df %>% filter(str_detect(pheno, "height") & alpha == "0.05"),
             aes(yintercept = threshold), linetype = 2) +
  geom_segment(data = pheno_peak_df %>% filter(str_detect(pheno, "height")),
               aes(x = ci_lo, xend = ci_hi, y = -0.1, yend = -0.1), 
               colour = "brown", size = 3) +
  facet_grid(nitrate ~ chrom, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(0, 500, 100)) +
  labs(x = "Position (cM)", title = "Height") +
  theme(panel.grid = element_blank())
```

```{r}
pheno_summary %>% 
  ggplot(aes(z68_shoots_mean_ln, senescence_height_mean_ln)) +
  geom_point()
```


## PCA

```{r}
pca_data <- ril_phenotypes %>% 
  filter(str_detect(ril_id, "RIL")) %>% 
  mutate(senescence_height = ifelse(identifier == "B5-HN-079", NA, senescence_height),
         z68_shoots = ifelse(identifier == "B6-HN-087", NA, z68_shoots)) %>% 
  select(batch, nitrate_level, ril_id, 
         z68_shoots, flowering_time, z68_greenness, senescence_height, 
         seed_weight_g, lifespan, flowering_senescence_interval) %>% 
  drop_na()

pca_result <- pca_data %>% 
  select(-batch, -nitrate_level, -ril_id) %>% 
  prcomp(scale = TRUE)

library(ggfortify)
autoplot(pca_result, data = pca_data,
         colour = "nitrate_level", loadings = TRUE, loadings.label = TRUE)
```

Let's see eigenvalues

```{r}
pca_result %>% 
  broom::tidy("pcs") %>% 
  ggplot(aes(PC, percent)) +
  geom_col() + geom_line(aes(y = cumulative))
```


Let's see where the loadings are:

```{r}
pca_result %>% 
  tidy(matrix = "variables") %>% 
  filter(PC %in% 1:4) %>% 
  ggplot(aes(value, column)) +
  geom_point() +
  geom_vline(xintercept = 0) +
  facet_wrap(~ PC) +
  theme(panel.grid = element_blank())
```


```{r}
pca_summary <- pca_result$x %>% 
  as_tibble() %>% 
  select(PC1, PC2, PC3, PC4) %>% 
  bind_cols(pca_data) %>% 
  group_by(nitrate_level, ril_id) %>% 
  summarise(pc1_mean = mean(PC1),
            pc2_mean = mean(PC2),
            pc3_mean = mean(PC3),
            pc4_mean = mean(PC4)) %>% 
  pivot_wider(names_from = "nitrate_level", values_from = matches("pc")) %>% 
  mutate(pc1_mean_PLAS = pc1_mean_HN - pc1_mean_LN,
         pc2_mean_PLAS = pc2_mean_HN - pc2_mean_LN,
         pc3_mean_PLAS = pc3_mean_HN - pc3_mean_LN,
         pc4_mean_PLAS = pc4_mean_HN - pc4_mean_LN) %>% 
  column_to_rownames("ril_id") %>% 
  as.matrix()

pca_scan <- scan1(geno_prob, pca_summary, cores = 2) %>% 
  tidy(map = geno_map) %>% 
  mutate(nitrate = str_remove(pheno, ".*_"),
         pheno = str_remove(pheno, "_.*"))
```

```{r}
pca_scan %>% 
  ggplot(aes(pos, LOD)) +
  geom_line(aes(colour = pheno)) +
  geom_hline(yintercept = 3, linetype = 2) +
  facet_grid(nitrate ~ chrom, scales = "free_x", space = "free_x") +
  theme(panel.grid = element_blank()) +
  scale_colour_viridis_d()
```

